src/Chainweb/Pact/Backend/InMemoryCheckpointer.hs

```diff
@@ -13,59 +13,112 @@ module Chainweb.Pact.Backend.InMemoryCheckpointer
+--         Just (height, _) -> modifyMVar_ lock (return . HMS.insert (Just (height, hash)) p)
+--   return result
+--   where
+--     restore' toput = withMVarMasked  lock $ \store -> do
+--       case HMS.lookup mHeightHash store of
+--         Just pactdbenv -> do
+--           putMVar toput pactdbenv
+--           return $ PactDbEnv' pactdbenv
+--         -- TODO
+--         Nothing -> internalError $ case mHeightHash of
+--           Just (bh, parentHash) ->
+--               "InMemoryCheckpointer: Restore not found: height=" <> pack (show bh)
+--               <> ", hash=" <> pack (show parentHash)
+--               <> ", known=" <> pack (show (HMS.keys store))
+--           Nothing -> "Something went wrong at genesis. There is nothing here!"
+
 
-initInMemoryCheckpointEnv :: P.Logger -> P.GasEnv -> IO CheckpointEnv
-initInMemoryCheckpointEnv logger gasEnv = do
-    inmem <- newMVar mempty
+
+data Store = Store
+  { _theStore :: HashMap (BlockHeight, BlockHash) (DbEnv PureDb)
+  , _tempSaveHeight :: Maybe BlockHeight
+  , _dbenv :: MVar (DbEnv PureDb)
+  }
+
+initInMemoryCheckpointEnv :: Loggers -> Logger -> GasEnv -> IO (PactDbEnv', CheckpointEnv)
+initInMemoryCheckpointEnv loggers logger gasEnv = do
+    pdenv@(PactDbEnv _ env) <- mkPureEnv loggers
+    initSchema pdenv
+    inmem <- newMVar (Store mempty Nothing env)
     return $
-        CheckpointEnv
+        (PactDbEnv' pdenv,
+          CheckpointEnv
             { _cpeCheckpointer =
-                  Checkpointer
-                      { restore = restore' inmem
-                      , restoreInitial = restoreInitial' inmem
-                      , save = save' inmem
-                      , saveInitial = saveInitial' inmem
-                      , discard = discard' inmem
-                      }
+                Checkpointer
+                    (doRestore inmem)
+                    -- (doRestoreInitial inmem)
+                    (doSave inmem)
+                    (doDiscard inmem)
             , _cpeLogger = logger
             , _cpeGasEnv = gasEnv
-            }
-
-type Store = HashMap (BlockHeight, BlockHash) PactDbState
+            })
 
-restore' :: MVar Store -> BlockHeight -> BlockHash -> IO (Either String PactDbState)
-restore' lock height hash = do
-    withMVarMasked lock $ \store -> do
-        case HMS.lookup (height, hash) store of
-            Just dbstate -> return (Right dbstate)
-            Nothing -> return $ Left $
-              "InMemoryCheckpointer: Restore not found: height=" <> show height
-              <> ", hash=" <> show hash
-              <> ", known=" <> show (HMS.keys store)
+doRestore :: MVar Store ->  Maybe (BlockHeight, ParentHash) -> IO PactDbEnv'
+doRestore lock (Just (height, hash)) = do
+    pactdbenv <- withMVarMasked lock $ \store -> do
+        case HMS.lookup (pred height, hash) (_theStore store) of
+            Just dbenv -> do
+              mdbenv <- newMVar dbenv
```

Comment:
you need to store `mdbenv` in `_dbenv` here, for retrieval in `save` (and `discard` should probably reset it, so maybe `_dbenv` should be `Maybe (MVar (DbEnv PureDb))`

Reply:
eh, skip the Maybe stuff. I think if you store it here that should fix it